log.go

```diff
@@ -382,18 +382,22 @@ func (l *Logger) Err(err error) *Event {
 	return l.Info()
 }
 
-// Fatal starts a new message with fatal level. The os.Exit(1) function
-// is called by the Msg method, which terminates the program immediately.
+// Fatal starts a new message with fatal level. The ExitFunc(1) function
+// (os.Exit by default) is called by the Msg method, which terminates the 
+// program immediately.
 //
 // You must call Msg on the returned event in order to send the event.
 func (l *Logger) Fatal() *Event {
 	return l.newEvent(FatalLevel, func(msg string) {
 		if closer, ok := l.w.(io.Closer); ok {
 			// Close the writer to flush any buffered message. Otherwise the message
-			// will be lost as os.Exit() terminates the program immediately.
+			// will be lost as ExitFunc() (os.Exit by default) terminates the program immediately.
 			closer.Close()
 		}
-		os.Exit(1)
+		if ExitFunc == nil {
+			ExitFunc = os.Exit
+		}
+		ExitFunc(1)
 	})
 }
```

Comment:
About the problem I raised [here](https://github.com/rs/zerolog/pull/717#discussion_r2006423308), here is what I suggest you to do

```go
type ExitFunc func(int)

type Logger struct {
        w       LevelWriter
        level   Level
        sampler Sampler
        context []byte
        hooks   []Hook
        stack   bool
        ctx     context.Context
        exiter ExitFunc
}
```

```suggestion
		exitFunc = l.exitFunc
		if exitFunc == nil {
			exitFunc = os.Exit
		}
		exitFunc(1)
 	})
 }

func(l *logger) SetTestExiter(t *testing.T, exitFunc ExitFunc) {
	l.exitFunc = exitFunc
	t.Cleanup(func() {
		l.exitFunc = nil
	})
}
```

This is a safe way, there won't be data race, no global overlapping between packages calling module

Reply:
super valid criticisms, thanks for the feedback

Reply:
for the setter, it seems more inline with other functions to have a function such as the following, to return a new child Logger instance with the desired ExitFunc than to modify it?

```go
// ExitFunc returns a logger with the e ExitFunc.
func (l Logger) ExitFunc(e ExitFunc) Logger {
	l.exiter = e
	return l
}
```

the intention wasn't to make this functionality exclusively available for testing purposes so I am not sure I see a reason to add the suggest `SetTestExiter`? Would having both make sense in your eyes?

Reply:
I have pushed some modifications based on your feedback. For the time being I have chosen to not tie the setting of `l.exitFunc` to testing and will wait for additional feedback from others.